libgadget/blackhole.c

```diff
@@ -599,40 +945,73 @@ blackhole_accretion_ngbiter(TreeWalkQueryBHAccretion * I,
             }
         }
     }
+    
+    
+    /********************************************/
+    /* Find min pot pos within halo */
+    if(r < 10)
+    {
+        if(P[other].Potential < O->BH_HaloMinPot)
+        {
+            int d;
+            O->BH_HaloMinPot = P[other].Potential;
+            for(d = 0; d < 3; d++) {
+                O->BH_HaloMinPotPos[d] = P[other].Pos[d];
+            }
+        }
+    }
+    /********************************************/
 
     /* Accretion / merger doesn't do self interaction */
     if(P[other].ID == I->ID) return;
-
-    if(P[other].Type == 5 && r2 < iter->accretion_kernel.HH)	/* we have a black hole merger */
+    
+    /* we have a black hole merger, use 2 times GravitationalSoftening as merging criteria*/
+    if(P[other].Type == 5 && r < (2*FORCE_SOFTENING(0,1)/2.8))	
     {
-        /* We do not depend on the BH relative velocity.
-         * Because the BHs are not dissipative, their relative velocities
-         * can be large, causing clumps of BHs to build up
-         * at the same position without merging. */
-        MyIDType readid, newswallowid;
-
-        #pragma omp atomic read
-        readid = (BHP(other).SwallowID);
-
-        /* Here we mark the black hole as "ready to be swallowed" using the SwallowID.
-         * The actual swallowing is done in the feedback treewalk by setting Swallowed = 1
-         * and merging the masses.*/
-        do {
-            /* Generate the new ID from the old*/
-            if(readid != (MyIDType) -1 && readid < I->ID ) {
-                /* Already marked, prefer to be swallowed by a bigger ID */
-                newswallowid = I->ID;
-            } else if(readid == (MyIDType) -1 && P[other].ID < I->ID) {
-                /* Unmarked, the BH with bigger ID swallows */
-                newswallowid = I->ID;
-            }
-            else
-                break;
-        /* Swap in the new id only if the old one hasn't changed:
-         * in principle an extension, but supported on at least clang >= 9, gcc >= 5 and icc >= 18.*/
-        } while(!__atomic_compare_exchange_n(&(BHP(other).SwallowID), &readid, newswallowid, 0, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
+        O->encounter = 1; // mark the event when two BHs encounter each other
+        
+        int d;
+        double KE = 0;
+        double PE = 0;
+        
+        for(d = 0; d < 3; d++){
+            KE += 0.5 * pow(I->Vel[d] - P[other].Vel[d], 2);
+            PE += (I->Accel[d] - P[other].GravAccel[d]) * (I->Pos[d] - P[other].Pos[d]);
+        }
+        
+        KE /= (All.cf.a*All.cf.a); /* convert to proper velocity */
+        PE /= All.cf.a; /* convert to proper unit */
+        
+        /* add option to merge the BHs if they are gravitationally bounded */
+        if(blackhole_params.BlackHoleRepositionEnabled == 1 || blackhole_params.MergeGravBound == 0 || (PE + KE < 0 && blackhole_params.MergeGravBound == 1))
```

Comment:
The "blackhole_params.MergeGravBound == 0" is redundant

Reply:
I would like to keep the option to turn off the Gravitational Bound criteria.

Reply:
I see, I misread the conditional. Could you reword it to make it a bit shorter?
Maybe pull this code out to a short function that computes the PE and KE and checks if they are gravitationally bound?